1.
E0602:Undefined variable 'SQLAlchemy'

Cause:  commented out the following line

from flask_sqlalchemy import SQLAlchemy

2.
E0602:Undefined variable 'text'

Cause:  commented out the following line

from sqlalchemy.sql import text

which affects the following function:

def query_db(self, query, data=None):
        result = self.db.session.execute(text(query), data)

3.
E0602:Undefined variable 'db'

Cause:  mistype a variable name

def __init__(self, app, d):
        config = {
                'host': 'localhost',
                'database': db, # we got db as an argument
                'user': 'root',
                'password': 'root',
                'port': '3306' # change the port to match the port your SQL server is running on
        }

4. 
E0001:invalid syntax (<string>, line 29)

Cause:  mistyped if statement

elf query[0:6].lower() == 'insert':

5.
sqlalchemy.exc.StatementError
StatementError: (sqlalchemy.exc.InvalidRequestError) 
A value is required for bind parameter 'specific_id' [SQL: u'SELECT * FROM users WHERE user.id = %s;'] 
[parameters: [{}]] (Background on this error at: http://sqlalche.me/e/cd3x)

Cause:  the query and data dictionaries have to be created, see that syntax complies with MySQL Workbench standards

6. sqlalchemy.exc.OperationalError
OperationalError: (_mysql_exceptions.OperationalError) 
(1054, "Unknown column 'user.id' in 'where clause'") [SQL: u'SELECT * FROM users WHERE user.id = %s;'] 
[parameters: (u'1',)] (Background on this error at: http://sqlalche.me/e/e3q8) 

Cause:  the table is called 'users' and not 'user', so correct syntax to read 'users.id'

7.
sqlalchemy.exc.ProgrammingError
ProgrammingError: (_mysql_exceptions.ProgrammingError) (1064, "You have an error in your SQL syntax; 
check the manual that corresponds to your MySQL server version for the right syntax to use near 
'UDATE users SET users.email = 'jacksonfive@jacksonfive.com', users.password = 'm' at line 1") 
[SQL: u'UDATE users SET users.email = %s, users.password = %s'] 
[parameters: (u'jacksonfive@jacksonfive.com', u'michaelrocks')] (Background on this error at: http://sqlalche.me/e/f405)

Cause:  syntax error inside the query, check syntax and retry

8.
werkzeug.exceptions.BadRequestKeyError
BadRequestKeyError: 400 Bad Request: KeyError: 'user_id'

Cause:  the data on the html is not passing through the handler function; be sure that the syntax matches all the way 
from the html to the link arguments and variables in side the function

9.
Method Not Allowed
The method is not allowed for the requested URL.

Cause:  no method specified on Flask route, and posting from form, so need to specify whether POST or GET.

10.
Not Found
The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.

Cause:  the route not found anywhere on flask.  check syntax of html <form action='#####'> and flask routes, and correct where necessary.

11.
sqlalchemy.exc.OperationalError
OperationalError: (_mysql_exceptions.OperationalError) 
(1049, "Unknown database 'email_validations'") 
(Background on this error at: http://sqlalche.me/e/e3q8)

Cause:  the database with that name does not exist in MySQL.  Check if it does and its correct name.

12.
UnboundLocalError
UnboundLocalError: local variable 'email' referenced before assignment

Cause:  python reads from top to bottom.  make sure you declare variables before making use of them.

13.
TypeError
TypeError: The view function did not return a valid response. The function either returned None or ended without a return statement.

Cause:  no return statement in the function.  make sure to return something at end of handler function

14.
ValueError
ValueError: too many values to unpack

Cause: got this message when trying to implement flask flash categories
Fix: include the following inside the get_flashed_messages() function:

 {% with messages = get_flashed_messages(with_categories=true) %}

